File: StudentskiPosao/app/src/main/java/hr/firma/sp/studentskiposao/algorithm/RedBlackBST.java
package hr.firma.sp.studentskiposao.algorithm;

// isto ko u knjizi - Key i Value su tipovi koje zadajes prilikom kreiranja objekta RedBlackBST
public class RedBlackBST<Key extends Comparable<Key>, Value> {

    private static final boolean RED   = true;
    private static final boolean BLACK = false;

    private Node<Key, Value> root;

    // primjeti, u knjizi je h.color (sintaksa C# jezika), ovdje je razlika samo u sintaksi - h.getColor()
    // slicno u nastavcima set() get()
    private boolean isRed(Node<Key, Value> h) {
        if (h == null) return false;
        return h.getColor() == RED;
    }

    private Node<Key, Value> rotateLeft(Node<Key, Value> h) {
        Node<Key, Value> x = h.getRight();
        h.setRight(x.getLeft());
        x.setLeft(h);
        x.setColor(h.getColor());
        h.setColor(RED);
        x.setN(h.getN());
        h.setN(1 + size(h.getLeft()) + size(h.getRight()));

        return x;
    }

    private Node<Key, Value> rotateRight(Node<Key, Value> h) {
        Node<Key, Value> x = h.getLeft();
        h.setLeft(x.getRight());
        x.setRight(h);
        x.setColor(h.getColor());
        h.setColor(RED);
        x.setN(h.getN());
        h.setN(1 + size(h.getLeft()) + size(h.getRight()));

        return x;
    }

    private void flipColors(Node<Key, Value> h) {
        h.setColor(RED);
        h.getLeft().setColor(BLACK);
        h.getRight().setColor(BLACK);
    }

    private int size(Node<Key, Value> x) {
        if (x == null) return 0;
        else return x.getN();
    }

    private Node<Key, Value> put(Node<Key, Value> h, Key key, Value val) {
        if (h == null) return new Node(key, val, 1, RED);

        int cmp = key.compareTo(h.getKey());

        if (cmp < 0) h.setLeft(put(h.getLeft(), key, val));
        else if (cmp > 0) h.setRight(put(h.getRight(), key, val));
        else h.setVal(val);

        if (isRed(h.getRight()) && !isRed(h.getLeft())) h = rotateLeft(h);
        if (isRed(h.getLeft()) && isRed(h.getLeft().getLeft())) h = rotateRight(h);
        if (isRed(h.getLeft()) && isRed(h.getRight())) flipColors(h);

        h.setN(size(h.getLeft()) + size(h.getRight()) + 1);

        return h;
    }

    // ova metoda opisana pod binary search tree obicnim cini mi se, ne podredblackbst, ali je ista za oba algoritma
    private Value get(Node<Key,  Value> h, Key key) {
        if (h == null) return null;
        int cmp = key.compareTo(h.getKey());
        if (cmp < 0) return get(h.getLeft(), key);
        else if (cmp > 0) return get(h.getRight(),  key);
        else return h.getVal();
    }

    public int size() {
        return size(root);
    }

    public void put(Key key, Value val) {
        root = put(root, key, val);
        root.setColor(RED);
    }

    public Value get(Key key) {
        return get(root, key);
    }

}

File: StudentskiPosao/app/src/main/java/hr/firma/sp/studentskiposao/algorithm/RedBlackBST/RedBlackBST.kt
package hr.firma.sp.studentskiposao.algorithm.RedBlackBST

object Colors {
    val RED = true
    val BLACK = false
}

class RedBlackBST<Key : Comparable<Key>, Value> {

    private var root: Node<Key, Value>? = null

    private fun isRed(h: Node<Key, Value>?) : Boolean {
        if (h == null) return false
        return h.color!!
    }

    private fun rotateLeft(h: Node<Key,  Value>) : Node<Key, Value> {

        var x = h.right!!
        h.right = x.left
        x.left = h
        x.color = h.color
        h.color = Colors.RED
        x.N = h.N
        h.N = 1 + size(h.left) + size(h.right)

        return x

    }

    private fun rotateRight(h: Node<Key, Value>) : Node<Key, Value> {

        var x = h.left!!
        h.left = x.right
        x.right = h
        x.color = h.color
        h.color = Colors.RED
        x.N = h.N
        h.N = 1 + size(h.left) + size(h.right)

        return x
    }

    private fun flipColors(h: Node<Key, Value>) {

        h.color =  Colors.RED
        h.left?.color = Colors.BLACK
        h.right?.color = Colors.BLACK

    }

    private fun size(x: Node<Key, Value>?) : Int {
        if (x == null) return 0
        else return x.N
    }

    private fun put(h: Node<Key, Value>?, key: Key, value: Value) : Node<Key, Value> {

        if (h == null) return Node(key, value, 1, Colors.RED)

        var cmp: Int = key.compareTo(h.key!!)

        if (cmp < 0) h.left = put(h.left, key,  value)
        else if (cmp > 0) h.right = put(h.right, key, value)
        else h.value = value

        if (isRed(h.right) && !isRed(h.left)) h.setMe(rotateLeft(h))
        if (isRed(h.left) && isRed(h.left?.left)) h.setMe(rotateRight(h))
        if (isRed(h.left) && isRed(h.right)) flipColors(h)

        h.N = size(h.left) + size(h.right) + 1

        return h

    }

    private fun get(h: Node<Key, Value>?, key: Key) : Value? {
        if (h == null) return null
        var cmp = key.compareTo(h.key!!)
        if (cmp < 0) return get(h.left, key)
        else if (cmp > 0) return get(h.right, key)
        else return h.value
    }

    public fun size() : Int {
        return size(root)
    }

    public fun put(key: Key, value: Value) {
        root = put(root, key, value)
        root?.color = Colors.RED
    }

    public fun get(key: Key) : Value? {
        return get(root, key)
    }

}
